util/piggy_prepare.py

```diff
@@ -0,0 +1,339 @@
+import hashlib
+import inspect
+import json
+import os
+import sys
+import time
+
+
+def piggy_prepare_run(dir, script, source, target, is_pgxl):
+
+    date = time.strftime('%Y%m%d')
+    fp = date + '_' + hashlib.sha1('\t'.join([dir, script, source, target])).hexdigest()[:8]
+    fname = 'piggy_func_' + fp
+    sview = 'piggy_source_' + fp
+    tview = 'piggy_target_' + fp
+
+    if ' ' not in source:
+        source_query = 'SELECT * FROM ' + source
+    else:
+        source_query = source
+
+    if '(' not in target:
+        target_query = 'SELECT * FROM ' + target
+    else:
+        tname, tcols = target.strip(' )').split('(')
+        target_query = 'SELECT %s FROM %s' % (tcols, tname)
+
+    logfile_path = os.path.join(dir, fname + '.log')
+    if os.path.isfile(logfile_path):
+        os.remove(logfile_path)
+
+    ts_format = '%m-%d %H:%M:%S'
+
+    # Why TABLE not VIEW?
+    # Because PGXL does not create views on data nodes.
+
+    sql_create_tables = '''
+    DROP TABLE IF EXISTS %(sview)s CASCADE;
+    DROP TABLE IF EXISTS %(tview)s CASCADE;
+    CREATE TABLE %(sview)s AS SELECT * FROM (%(source_query)s) _sview_source LIMIT 0;
+    CREATE TABLE %(tview)s AS %(target_query)s LIMIT 0;
+    ''' % {
+        'sview': sview,
+        'tview': tview,
+        'source_query': source_query,
+        'target_query': target_query,
+    }
+
+
+    # Why log to file instead of a DB table?
+    # PGXL would complain something like 'no XID'.
+    # This may have fixed it though: https://github.com/snaga/postgres-xl/commit/d4136935a1d741c61ada13ef8c5ae44f68162cc9
+
+    sql_create_functions = '''
+    CREATE OR REPLACE FUNCTION %(fname)s (record %(sview)s)
+        RETURNS SETOF %(tview)s
+    AS $func$
+
+        # NULL input tuple indicates end of stream.
+        # It's time to flush output and clean up.
+        if record is None:
+            import os
+            if 'piggy' not in SD:
+                raise StopIteration
+            piggy = SD['piggy']
+            try:
+                piggy.log('Grabbing final tuples from the buffer.')
+                for x in piggy.pull_final():
+                    yield x
+            finally:
+                del SD['piggy']
+            raise StopIteration
+
+        # First input tuple; we setup the worker process.
+        if 'piggy' not in SD:
+            import errno
+            import fcntl
+            import json
+            import os
+            import socket
+            import subprocess
+            import time
+            from datetime import datetime
+            from Queue import Queue, Empty, Full
+            from threading import Thread
+
+            # Create a pipe for logs.
+            # Only one process would succeed on the same host.
+            log_pipe_name = os.path.join('%(dir)s', '%(fname)s.pipe')
+            try:
+                os.mkfifo(log_pipe_name)
```

Comment:
Never mind. Found the issue -- DD was creating a new DB connection every time we call get_log, and the shared dict in play does not go beyond one process / connection, so the sucker thread became an orphan, and subsequently calls try to create more sucker threads.

Direct SSH access would definitely make things much easier, but the downside with it is that 1) the DB server / cluster doesn't always allows ssh access to all DD developers who'd like to use the DB; 2) we'd have to manage the SSH keys / passwords in addition to the JDBC params in the DD client. Longer term we may well have SSH access as a deployment requirement. I'd be as happy as you will be to switch to ssh when that day comes :)